File: DSA/Linkedlist/Code/DeletionInDoublyLinkedList.java
// Deletion in Doubly Linked List in Java

class Node {
    int data;
    Node next;
    Node prev;
    
    Node(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

public class DoublyLinkedListDeletion {
    Node head;
    
    void addNode(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) temp = temp.next;
        temp.next = newNode;
        newNode.prev = temp;
    }
    
    void deleteAtPosition(int pos) {
        if (head == null) return;
        
        if (pos == 1) {
            head = head.next;
            if (head != null)
                head.prev = null;
            return;
        }
        
        Node temp = head;
        for (int i = 1; i < pos && temp != null; i++)
            temp = temp.next;
        
        if (temp == null) return;
        
        if (temp.next != null)
            temp.next.prev = temp.prev;
        if (temp.prev != null)
            temp.prev.next = temp.next;
    }
    
    void deleteByValue(int value) {
        if (head == null) return;
        
        if (head.data == value) {
            head = head.next;
            if (head != null)
                head.prev = null;
            return;
        }
        
        Node temp = head;
        while (temp != null && temp.data != value)
            temp = temp.next;
        
        if (temp == null) return;
        
        if (temp.next != null)
            temp.next.prev = temp.prev;
        if (temp.prev != null)
            temp.prev.next = temp.next;
    }
    
    void displayForward() {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " <-> ");
            temp = temp.next;
        }
        System.out.println("NULL");
    }
    
    public static void main(String[] args) {
        DoublyLinkedListDeletion list = new DoublyLinkedListDeletion();
        
        System.out.println("=== Deletion in Doubly Linked List ===\n");
        
        // Build list
        for (int i = 1; i <= 5; i++)
            list.addNode(i * 10);
        
        System.out.print("Original list: ");
        list.displayForward();
        
        list.deleteAtPosition(2);
        System.out.print("After deleting position 2: ");
        list.displayForward();
        
        list.deleteByValue(40);
        System.out.print("After deleting value 40: ");
        list.displayForward();
        
        System.out.println("\nComplexity Analysis:");
        System.out.println("Delete at position: O(n)");
        System.out.println("Delete by value: O(n)");
        System.out.println("Space: O(1)");
    }
}

File: DSA/Linkedlist/Code/DeletionInDoublyLinkedList.py
# Deletion in Doubly Linked List in Python

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLinkedListDeletion:
    def __init__(self):
        self.head = None
    
    def add_node(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        temp.next = new_node
        new_node.prev = temp
    
    def delete_at_position(self, pos):
        if self.head is None:
            return
        
        if pos == 1:
            self.head = self.head.next
            if self.head is not None:
                self.head.prev = None
            return
        
        temp = self.head
        for i in range(1, pos):
            if temp is None:
                return
            temp = temp.next
        
        if temp is None:
            return
        
        if temp.next is not None:
            temp.next.prev = temp.prev
        if temp.prev is not None:
            temp.prev.next = temp.next
    
    def delete_by_value(self, value):
        if self.head is None:
            return
        
        if self.head.data == value:
            self.head = self.head.next
            if self.head is not None:
                self.head.prev = None
            return
        
        temp = self.head
        while temp is not None and temp.data != value:
            temp = temp.next
        
        if temp is None:
            return
        
        if temp.next is not None:
            temp.next.prev = temp.prev
        if temp.prev is not None:
            temp.prev.next = temp.next
    
    def display_forward(self):
        temp = self.head
        elements = []
        while temp is not None:
            elements.append(str(temp.data))
            temp = temp.next
        print(" <-> ".join(elements) + " <-> NULL")

# Test
list1 = DoublyLinkedListDeletion()

print("=== Deletion in Doubly Linked List ===\n")

# Build list
for i in range(1, 6):
    list1.add_node(i * 10)

print("Original list: ", end="")
list1.display_forward()

list1.delete_at_position(2)
print("After deleting position 2: ", end="")
list1.display_forward()

list1.delete_by_value(40)
print("After deleting value 40: ", end="")
list1.display_forward()

print("\nComplexity Analysis:")
print("Delete at position: O(n)")
print("Delete by value: O(n)")
print("Space: O(1)")
